src/porepy/numerics/fv/tpfa.py

```diff
@@ -133,8 +133,39 @@ def discretize(self, g, data):
         t_b[is_neu] = 1
         t_b = t_b[bndr_ind]
         t[is_neu] = 0
+
+        # Calculate transmissibilities for periodic boundary faces.
+        # Left faces are periodic with right faces
+        left = bnd.per_map[0]
+        right = bnd.per_map[1]
+        t_per = np.zeros(g.num_faces)
+        t_per[left] = t[left] * t[right] / (t[left] + t[right])
+        t_per[right] = t_per[left]
+        t[bnd.is_per] = t_per[bnd.is_per]
+
+        # Add connection between the left cells and right faces (and vice versa).
+        # The flux over the periodic boundary face is defined equivalently to the
+        # flux over an internal face: flux_left = T_left * (p_left - p_right).
+        # The term T_left * p_left is already included in fi and ci, but we need
+        # to add the second term T_left * (-p_right). Equivalently for flux_right.
+        left_fi, left_ci, left_sgn = sps.find(g.cell_faces[left])
+        right_fi, right_ci, right_sgn = sps.find(g.cell_faces[right])
+        # Sort subface indices to not loose left to right mapping
+        I_left = np.argsort(left_fi)
+        I_right = np.argsort(right_fi)
+        if not (np.array_equal(left_fi[I_left], np.arange(left.size)) and
+                np.array_equal(right_fi[I_right], np.arange(right.size))):
+            raise RuntimeError("Could not find correct periodic boundary mapping")
+        left_ci = left_ci[I_left]
+        right_ci = right_ci[I_right]
+
+        # Add periodic boundaries to subface cell mapping
+        face_idx = np.hstack((fi, left, right))
+        cell_idx = np.hstack((ci, right_ci, left_ci))
+        sgn_face = np.hstack((sgn, -left_sgn, -right_sgn))
+
         # Create flux matrix
-        flux = sps.coo_matrix((t[fi] * sgn, (fi, ci))).tocsr()
+        flux = sps.coo_matrix((t[face_idx] * sgn_face, (face_idx, cell_idx))).tocsr()
```

Comment:
How about vector sources? I know this is not a primary concern for you, but I am worried about fragmentation of the code. I think the extension should be straightforward - would you care to have a short look and see what it would take?

Reply:
I had a look at vector sources and I believe they are included now. But I could not find any unittests for vector sources so it is a bit hard to know for sure. Maybe you can point me to the correct test?

Reply:
unit/test_mpfa_gravity.py. If you have a suggestion for a better name for the file, I am listening.

Reply:
Maybe test_fv_vector_source.py ?